lib/engine/game/g_1829/step/buy_sell_par_shares.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+require_relative '../../../step/buy_sell_par_shares'
+
+module Engine
+  module Game
+    module G1829
+      module Step
+        class BuySellParShares < Engine::Step::BuySellParShares
+          FIRST_SR_ACTIONS = %w[buy_company pass].freeze
+          EXCHANGE_ACTIONS = %w[buy_shares pass].freeze
+
+          def actions(entity)
+            return [] unless entity&.player?
+
+            result = super
+
+            result.concat(FIRST_SR_ACTIONS) if can_buy_company?(entity)
+            result.concat(EXCHANGE_ACTIONS) if can_exchange?(entity)
+            result
+          end
+
+          def can_buy_company?(player, company = nil)
+            return false if first_sr_passed?(player) || @game.num_certs(player) >= @game.cert_limit
+            return buyable_company?(player, company) if company
+
+            @game.buyable_bank_owned_companies.any? { |c| buyable_company?(player, c) }
+          end
+
+          def buyable_company?(player, company)
+            return false if first_sr_passed?(player) || sold? || bought?
+            return false if @game.bond?(company) && @round.players_sold[player][:bond]
+
+            player.cash >= company.value
+          end
+
+          def buyable_bank_owned_companies
+            @companies.select { |c| c.owner == @bank && c.revenue > 20 }
+          end
+
+          def available
+            @companies.select { |c| !c.closed? && c.owner == @bank && c.revenue > 20 }
```

Comment:
why don't you call buyable banked ownedcompanies in all these places/ you have too many repeated lines of code

Reply:
done

Reply:
there are many other references, please clean all of them up and fix the build